### PR TITLE
DRep undelegation fix

### LINKED ISSUE #4772: Unexpected removal of DRep delegation leading to an issue with withdrawal
An error was noticed when trying to submit a transaction including a withdrawal when the stake key has a vote delegation to a registered but inactive DRep. Multiple tests were performed. And it was not always the case when withdrawing rewards with a vote delegation to an inactive DRep that the error was given. 

Describing the flow of events to make it a bit clearer. 

1. Issue noticed and multiple re-delegations and transactions where performed. Rewards was withdrawn in one of the tests. 
2. To re-construct the issue in a more formal way, the wallet was vote delegated to DRep with hex `2274519ac5359c003a7ace69c475ae55a86eb8f9fec6cf6feaada1debf`. This DRep was registered in epoch `724`, updated in `734` and marked as inactive in `754`. Rewards was successfully withdrawn in epoch `758`, tx hash `de2342e28817d3117e8fd09e01dd948668cb15a61d1cfcdf914eb8861c9d404f`. This confirmed how it should work, that inactive state should not prevent withdrawal.
3. Not convinced that the result in (2) was the full story we found the DRep delegation we originally had when we got the error and re-delegated vote power to this in epoch `758` after test in (2), hex `22fa6a8dc2635dddcf9af495cb144f7eb4ff845866fe48695ad7cb65d3`. This DRep is from own account but doesn't matter. Registered in epoch `664` and vote cast in `684` as the only action. Thus expired in epoch `704`.
4. We waited until epoch `759` to receive new rewards to withdraw as it was easier with the tools at hand (I know you can withdraw 0 lovelace). Now we get the error message again as we originally found. Tried both w/ and w/o 258 tagged sets to be sure it wasn't connected to that.

Network: `Preview Testnet`
Abs slot for submission test: `65610177`
Node version: `cardano-node 10.1.1 - linux-x86_64 - ghc-8.10 git rev 4184f9297bf7306713bfbb8f39eef61c056198cc`
CLI version: `cardano-cli 10.1.1.0 - linux-x86_64 - ghc-9.6 git rev 0000000000000000000000000000000000000000`
Ogmios version: `v6.9.0 (a07c154e)`

Wallet address:
```
addr_test1qpfepft9zs3y8ejcv84tq6tkp00wdm46fr6h3am02leunk8dc55q34v2ggxw9hea4rr3rry933a2zdh60v43h237s8ks7t2dja
```

CBOR (w/o 258 tag)
```
84a60081825820efc7370f054274109615a8c0d1545b035ed8920df3965b688431a36b8e3839fa010182825839005390a565142243e65861eab069760bdee6eeba48f578f76f57f3c9d8edc52808d58a420ce2df3da8c7118c858c7aa136fa7b2b1baa3e81ed1a4bb25055825839005390a565142243e65861eab069760bdee6eeba48f578f76f57f3c9d8edc52808d58a420ce2df3da8c7118c858c7aa136fa7b2b1baa3e81ed1a004c4b40021a0002a9b9031a03e9468405a1581de0edc52808d58a420ce2df3da8c7118c858c7aa136fa7b2b1baa3e81ed1a0042018e0800a100828258209b5bd6d6a8142c69586493d367a3a561e26d76037ac3928939fdc374a110bbdf5840c238651cc86a703a3db55d9eebb31a4364a31efdec6bf16bf165ee2e2a9e6328ecaa04fec30411efe9c22ace476d5c558f49d0b05ed8bd1a82374e5edf3dd60c8258201cf1e6972f59c345fa295fab41261a1e2058d49b7441741dc711a9745f70b9df5840fdb77b2181578dfb55661546b1b7bde4b0c1bd8ef40c45516a9e5eb4132595d5bb79906bfc74bb95b3af403e9e800c83ea89744738d4b9ac98f2bf4c91760e05f5f6
```

CBOR (w/ 258 tag)
```
84a600d9010281825820efc7370f054274109615a8c0d1545b035ed8920df3965b688431a36b8e3839fa010182825839005390a565142243e65861eab069760bdee6eeba48f578f76f57f3c9d8edc52808d58a420ce2df3da8c7118c858c7aa136fa7b2b1baa3e81ed1a4bb24f4d825839005390a565142243e65861eab069760bdee6eeba48f578f76f57f3c9d8edc52808d58a420ce2df3da8c7118c858c7aa136fa7b2b1baa3e81ed1a004c4b40021a0002aac1031a03e94a6805a1581de0edc52808d58a420ce2df3da8c7118c858c7aa136fa7b2b1baa3e81ed1a0042018e0800a100d90102828258209b5bd6d6a8142c69586493d367a3a561e26d76037ac3928939fdc374a110bbdf5840802c5255c6c7e508378d6ca83484d0bf06c0f8ce6c08df0b10515af0e4f3caff728b84ee6edebcd3796ce7b4ae17012c8b92e9e8a23b72bc020799edb5cae70a8258201cf1e6972f59c345fa295fab41261a1e2058d49b7441741dc711a9745f70b9df58405074a4ccc1c0b8d5c5264b8501e3dfb9886998eba26857cc32b08bcf6f9f6bc294ef92e78d5456ddef8ab8268d8fdb79317d6015f993bdf3f30944c194e91c0df5f6
```

Ogmios error:
```json
{"jsonrpc":"2.0","method":"submitTransaction","error":{"code":3150,"message":"The transaction is attempting to withdraw rewards from stake credentials that do not engage in on-chain governance. Credentials must be associated with a delegate representative (registered, abstain or noConfidence) before associated rewards can be withdrawn. The field 'data.marginalizedCredentials' lists all the affected credentials.","data":{"marginalizedCredentials":["edc52808d58a420ce2df3da8c7118c858c7aa136fa7b2b1baa3e81ed"]}},"id":null}
```

Node error:
```
Command failed: transaction submit  Error: Error while submitting tx: ShelleyTxValidationError ShelleyBasedEraConway (ApplyTxError (ConwayWdrlNotDelegatedToDRep (KeyHash {unKeyHash = "edc52808d58a420ce2df3da8c7118c858c7aa136fa7b2b1baa3e81ed"} :| []) :| []))
```

### PR DESCRIPTION
**Pull request by lehins**
# Description

This PR fixes #4772 

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages.
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated.
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/HardFork.hs [hunk 1]
```diff
@@ -74,7 +74,11 @@ hardforkTransition = do
         epochState
           & esLStateL . lsCertStateL %~ \certState ->
             let umap = certState ^. certDStateL . dsUnifiedL
-                dReps = certState ^. certVStateL . vsDRepsL
+                dReps =
+                  -- Reset all delegations in order to remove any inconsistencies
+                  -- Delegations will be reset accordingly below.
+                  Map.map (\dRepState -> dRepState {drepDelegs = Set.empty}) $
+                    certState ^. certVStateL . vsDRepsL
```
Comment (neilmayhew) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/HardFork.hs:81]: There's a lens for `drepDelegs` so you could write this as

```haskell
Map.map (drepDelegsL .~ Set.empty) $
  certState ^. certVStateL . vsDRepsL
```

or even (if you want to be fancy)

```haskell
certState ^. certVStateL . vsDRepsL & traversed . drepDelegsL .~ Set.empty
```

However, no problem leaving it as it is if you want to get this finished sooner.

**Review by neilmayhew — APPROVED** [review]
LGTM

Nice use of TDD :-)

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/HardFork.hs (hunk 1 above)
Comment (lehins) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/HardFork.hs:81]: I'll leave it as-is in order to speed up the backport release process


> or even (if you want to be fancy)

We should definitely avoid being fancy! Using `traversed` here is a totally cryptic lensy way of doing a very simple thing.

**Review by Crypto2099 — COMMENTED** [review]
LGTM 😄

**Pull request merged**